File: 0046-permutations/0046-permutations.rb
# @param {Integer[]} nums
# @return {Integer[][]}

$res = []
def backtrack(nums, temp, visited)
    if(temp.size == nums.size)
        $res << temp.dup
        return
    end
    
    n = nums.size
    (0..n-1).each do |i|
        if(visited.include?(nums[i]) == false)
            visited << nums[i]
            temp << nums[i]
            backtrack(nums, temp, visited)
            temp.pop
            visited.delete(nums[i])
        end
    end
    
end


def permute(nums)
    temp=[]
    visited = Set.new
    backtrack(nums, temp, visited)
    ans = $res
    $res = []
    ans
end

File: 0046-permutations/0046-permutations.cpp
class Solution {
private:
    void backtrack(int i, vector<int>& nums, vector<int>& temp, unordered_map<int, bool>& visited, vector<vector<int>>& res){
        
        if(temp.size()==nums.size()){
            res.push_back(temp);
            return;
        }
        
        
        for(int k=0;k<nums.size();k++){
            if(!visited[nums[k]]){
                visited[nums[k]] = true;
                temp.push_back(nums[k]);
                backtrack(k, nums, temp, visited, res);
                temp.pop_back();
                visited[nums[k]]=false;
            }
        }
        
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        unordered_map<int, bool> visited;
        for(int i=-10;i<=10;i++) visited[i]=false;
        vector<int> temp;
        vector<vector<int>> res;
        backtrack(0, nums, temp, visited, res);
        return res;
    }
};